crypto/fipsmodule/hmac/hmac.c

```diff
@@ -64,34 +64,130 @@
+    out->methods[idx].update = AWS_LC_TRAMPOLINE_##HASH_NAME##_Update;  \
+    out->methods[idx].finalize = AWS_LC_TRAMPOLINE_##HASH_NAME##_Final; \
+    idx++;                                                              \
+    assert(idx <= HMAC_METHOD_MAX);                                     \
+  }
+
+DEFINE_LOCAL_DATA(struct hmac_method_array_st, AWSLC_hmac_in_place_methods) {
+  OPENSSL_memset((void*) out->methods, 0, sizeof(out->methods));
+  int idx = 0;
+  DEFINE_IN_PLACE_METHODS(EVP_md4(), MD4);
+  DEFINE_IN_PLACE_METHODS(EVP_md5(), MD5);
+  DEFINE_IN_PLACE_METHODS(EVP_sha1(), SHA1);
+  DEFINE_IN_PLACE_METHODS(EVP_sha224(), SHA224);
+  DEFINE_IN_PLACE_METHODS(EVP_sha256(), SHA256);
+  DEFINE_IN_PLACE_METHODS(EVP_sha384(), SHA384);
+  DEFINE_IN_PLACE_METHODS(EVP_sha512(), SHA512);
+  DEFINE_IN_PLACE_METHODS(EVP_sha512_256(), SHA512_256);
+}
+
+OPENSSL_STATIC_ASSERT(sizeof(((struct hmac_method_array_st *)NULL)->methods) <=
+                      HMAC_METHOD_MAX * sizeof(struct hmac_methods_st),
+                      hmac_methods_to_short_for_loop_limits_t)
+static const HmacMethods *GetInPlaceMethods(const EVP_MD *evp_md) {
+  const HmacMethods *methods = AWSLC_hmac_in_place_methods()->methods;
+  for (size_t idx = 0; idx < HMAC_METHOD_MAX; idx++) {
+    if (methods[idx].evp_md == evp_md) {
+      return &methods[idx];
+    }
+  }
+  return NULL;
+}
 
 uint8_t *HMAC(const EVP_MD *evp_md, const void *key, size_t key_len,
               const uint8_t *data, size_t data_len, uint8_t *out,
               unsigned int *out_len) {
-  // We have to verify that all the HMAC services actually succeed before
-  // updating the indicator state, so we lock the state here.
+  // We have to avoid the underlying SHA services updating the indicator
+  // state, so we lock the state here.
   FIPS_service_indicator_lock_state();
-  HMAC_CTX ctx;
-  HMAC_CTX_init(&ctx);
-  if (!HMAC_Init_ex(&ctx, key, key_len, evp_md, NULL) ||
-      !HMAC_Update(&ctx, data, data_len) ||
-      !HMAC_Final(&ctx, out, out_len)) {
-    out = NULL;
-  }
-  // Unlock service indicator state and check if HMAC functions have succeeded.
+  HMAC_CTX ctx = {.state = 0};
+  int result;
+  // If out is NULL, the digest is placed in a static array.
+  // Note: passing a NULL value for out to use the static array is not thread safe.
+  static uint8_t fallback_output[EVP_MAX_MD_SIZE] = {0};
+  if (out == NULL) {
+    out = fallback_output;
+  }
+
+  result = HMAC_Init_ex(&ctx, key, key_len, evp_md, NULL) &&
+           HMAC_Update(&ctx, data, data_len) &&
+           HMAC_Final(&ctx, out, out_len);
   FIPS_service_indicator_unlock_state();
-  if(out) {
+  HMAC_CTX_cleanup(&ctx);
+  if (result) {
     HMAC_verify_service_indicator(evp_md);
+    return out;
+  } else {
+    OPENSSL_cleanse(out, EVP_MD_size(evp_md));
+    return NULL;
   }
-  HMAC_CTX_cleanup(&ctx);
-  return out;
 }
 
 void HMAC_CTX_init(HMAC_CTX *ctx) {
```

Comment:
Do we want to initialise `state` with `HMAC_STATE_UNINITIALIZED`? 

Redundant now - but future-prove it a bit.

Reply:
Decided that `HMAC_STATE_UNINITIALIZED` must remain equal to `0` and added a comment about that.